File: qt/gen_qdirlisting.cpp
// +build ignore

#include <QDateTime>
#include <QDirListing>
#define WORKAROUND_INNER_CLASS_DEFINITION_QDirListing__DirEntry
#define WORKAROUND_INNER_CLASS_DEFINITION_QDirListing__const_iterator
#define WORKAROUND_INNER_CLASS_DEFINITION_QDirListing__sentinel
#include <QFileInfo>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QTimeZone>
#include <qdirlisting.h>
#include "gen_qdirlisting.h"

#ifndef _Bool
#define _Bool bool
#endif

QDirListing* QDirListing_new(struct miqt_string path) {
	QString path_QString = QString::fromUtf8(path.data, path.len);
	return new QDirListing(path_QString);
}

QDirListing* QDirListing_new2(struct miqt_string path, struct miqt_array /* of struct miqt_string */  nameFilters) {
	QString path_QString = QString::fromUtf8(path.data, path.len);
	QStringList nameFilters_QList;
	nameFilters_QList.reserve(nameFilters.len);
	struct miqt_string* nameFilters_arr = static_cast<struct miqt_string*>(nameFilters.data);
	for(size_t i = 0; i < nameFilters.len; ++i) {
		QString nameFilters_arr_i_QString = QString::fromUtf8(nameFilters_arr[i].data, nameFilters_arr[i].len);
		nameFilters_QList.push_back(nameFilters_arr_i_QString);
	}
	return new QDirListing(path_QString, nameFilters_QList);
}

QDirListing* QDirListing_new3(struct miqt_string path, IteratorFlags flags) {
	QString path_QString = QString::fromUtf8(path.data, path.len);
	return new QDirListing(path_QString, flags);
}

QDirListing* QDirListing_new4(struct miqt_string path, struct miqt_array /* of struct miqt_string */  nameFilters, IteratorFlags flags) {
	QString path_QString = QString::fromUtf8(path.data, path.len);
	QStringList nameFilters_QList;
	nameFilters_QList.reserve(nameFilters.len);
	struct miqt_string* nameFilters_arr = static_cast<struct miqt_string*>(nameFilters.data);
	for(size_t i = 0; i < nameFilters.len; ++i) {
		QString nameFilters_arr_i_QString = QString::fromUtf8(nameFilters_arr[i].data, nameFilters_arr[i].len);
		nameFilters_QList.push_back(nameFilters_arr_i_QString);
	}
	return new QDirListing(path_QString, nameFilters_QList, flags);
}

void QDirListing_Swap(QDirListing* self, QDirListing* other) {
	self->swap(*other);
}

struct miqt_string QDirListing_IteratorPath(const QDirListing* self) {
	QString _ret = self->iteratorPath();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

IteratorFlags QDirListing_IteratorFlags(const QDirListing* self) {
	return self->iteratorFlags();
}

struct miqt_array /* of struct miqt_string */  QDirListing_NameFilters(const QDirListing* self) {
	QStringList _ret = self->nameFilters();
	// Convert QList<> from C++ memory to manually-managed C memory
	struct miqt_string* _arr = static_cast<struct miqt_string*>(malloc(sizeof(struct miqt_string) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		QString _lv_ret = _ret[i];
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray _lv_b = _lv_ret.toUtf8();
		struct miqt_string _lv_ms;
		_lv_ms.len = _lv_b.length();
		_lv_ms.data = static_cast<char*>(malloc(_lv_ms.len));
		memcpy(_lv_ms.data, _lv_b.data(), _lv_ms.len);
		_arr[i] = _lv_ms;
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

const_iterator QDirListing_Begin(const QDirListing* self) {
	return self->begin();
}

const_iterator QDirListing_Cbegin(const QDirListing* self) {
	return self->cbegin();
}

sentinel QDirListing_End(const QDirListing* self) {
	return self->end();
}

sentinel QDirListing_Cend(const QDirListing* self) {
	return self->cend();
}

const_iterator QDirListing_ConstBegin(const QDirListing* self) {
	return self->constBegin();
}

sentinel QDirListing_ConstEnd(const QDirListing* self) {
	return self->constEnd();
}

void QDirListing_Delete(QDirListing* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QDirListing*>( self );
	} else {
		delete self;
	}
}

QDirListing__DirEntry* QDirListing__DirEntry_new(const DirEntry* param1) {
	return new QDirListing::DirEntry(*param1);
}

QDirListing__DirEntry* QDirListing__DirEntry_new2() {
	return new QDirListing::DirEntry();
}

struct miqt_string QDirListing__DirEntry_FileName(const QDirListing__DirEntry* self) {
	QString _ret = self->fileName();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_BaseName(const QDirListing__DirEntry* self) {
	QString _ret = self->baseName();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_CompleteBaseName(const QDirListing__DirEntry* self) {
	QString _ret = self->completeBaseName();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_Suffix(const QDirListing__DirEntry* self) {
	QString _ret = self->suffix();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_BundleName(const QDirListing__DirEntry* self) {
	QString _ret = self->bundleName();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_CompleteSuffix(const QDirListing__DirEntry* self) {
	QString _ret = self->completeSuffix();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_FilePath(const QDirListing__DirEntry* self) {
	QString _ret = self->filePath();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QDirListing__DirEntry_IsDir(const QDirListing__DirEntry* self) {
	return self->isDir();
}

bool QDirListing__DirEntry_IsFile(const QDirListing__DirEntry* self) {
	return self->isFile();
}

bool QDirListing__DirEntry_IsSymLink(const QDirListing__DirEntry* self) {
	return self->isSymLink();
}

bool QDirListing__DirEntry_Exists(const QDirListing__DirEntry* self) {
	return self->exists();
}

bool QDirListing__DirEntry_IsHidden(const QDirListing__DirEntry* self) {
	return self->isHidden();
}

bool QDirListing__DirEntry_IsReadable(const QDirListing__DirEntry* self) {
	return self->isReadable();
}

bool QDirListing__DirEntry_IsWritable(const QDirListing__DirEntry* self) {
	return self->isWritable();
}

bool QDirListing__DirEntry_IsExecutable(const QDirListing__DirEntry* self) {
	return self->isExecutable();
}

QFileInfo* QDirListing__DirEntry_FileInfo(const QDirListing__DirEntry* self) {
	return new QFileInfo(self->fileInfo());
}

struct miqt_string QDirListing__DirEntry_CanonicalFilePath(const QDirListing__DirEntry* self) {
	QString _ret = self->canonicalFilePath();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_AbsoluteFilePath(const QDirListing__DirEntry* self) {
	QString _ret = self->absoluteFilePath();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDirListing__DirEntry_AbsolutePath(const QDirListing__DirEntry* self) {
	QString _ret = self->absolutePath();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

long long QDirListing__DirEntry_Size(const QDirListing__DirEntry* self) {
	qint64 _ret = self->size();
	return static_cast<long long>(_ret);
}

QDateTime* QDirListing__DirEntry_BirthTime(const QDirListing__DirEntry* self, QTimeZone* tz) {
	return new QDateTime(self->birthTime(*tz));
}

QDateTime* QDirListing__DirEntry_MetadataChangeTime(const QDirListing__DirEntry* self, QTimeZone* tz) {
	return new QDateTime(self->metadataChangeTime(*tz));
}

QDateTime* QDirListing__DirEntry_LastModified(const QDirListing__DirEntry* self, QTimeZone* tz) {
	return new QDateTime(self->lastModified(*tz));
}

QDateTime* QDirListing__DirEntry_LastRead(const QDirListing__DirEntry* self, QTimeZone* tz) {
	return new QDateTime(self->lastRead(*tz));
}

QDateTime* QDirListing__DirEntry_FileTime(const QDirListing__DirEntry* self, int typeVal, QTimeZone* tz) {
	return new QDateTime(self->fileTime(static_cast<QFileDevice::FileTime>(typeVal), *tz));
}

void QDirListing__DirEntry_OperatorAssign(QDirListing__DirEntry* self, const DirEntry* param1) {
	self->operator=(*param1);
}

void QDirListing__DirEntry_Delete(QDirListing__DirEntry* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QDirListing::DirEntry*>( self );
	} else {
		delete self;
	}
}

QDirListing__sentinel* QDirListing__sentinel_new() {
	return new QDirListing::sentinel();
}

QDirListing__sentinel* QDirListing__sentinel_new2(const sentinel* param1) {
	return new QDirListing::sentinel(*param1);
}

void QDirListing__sentinel_Delete(QDirListing__sentinel* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QDirListing::sentinel*>( self );
	} else {
		delete self;
	}
}

QDirListing__const_iterator* QDirListing__const_iterator_new() {
	return new QDirListing::const_iterator();
}

reference QDirListing__const_iterator_OperatorMultiply(const QDirListing__const_iterator* self) {
	return self->operator*();
}

pointer QDirListing__const_iterator_OperatorMinusGreater(const QDirListing__const_iterator* self) {
	return self->operator->();
}

const_iterator* QDirListing__const_iterator_OperatorPlusPlus(QDirListing__const_iterator* self) {
	return &self->operator++();
}

void QDirListing__const_iterator_OperatorPlusPlusWithInt(QDirListing__const_iterator* self, int param1) {
	self->operator++(static_cast<int>(param1));
}

void QDirListing__const_iterator_Delete(QDirListing__const_iterator* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QDirListing::const_iterator*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qdirlisting.go
package qt

import (
	"unsafe"
)

type QDirListing__IteratorFlag int

const (
	QDirListing__Default             QDirListing__IteratorFlag = 0
	QDirListing__ExcludeFiles        QDirListing__IteratorFlag = 4
	QDirListing__ExcludeDirs         QDirListing__IteratorFlag = 8
	QDirListing__ExcludeSpecial      QDirListing__IteratorFlag = 16
	QDirListing__ResolveSymlinks     QDirListing__IteratorFlag = 32
	QDirListing__FilesOnly           QDirListing__IteratorFlag = 24
	QDirListing__DirsOnly            QDirListing__IteratorFlag = 20
	QDirListing__IncludeHidden       QDirListing__IteratorFlag = 64
	QDirListing__IncludeDotAndDotDot QDirListing__IteratorFlag = 128
	QDirListing__CaseSensitive       QDirListing__IteratorFlag = 256
	QDirListing__Recursive           QDirListing__IteratorFlag = 1024
	QDirListing__FollowDirSymlinks   QDirListing__IteratorFlag = 2048
)

type QDirListing struct {
	h          uintptr
	isSubclass bool
}

// NewQDirListing constructs a new QDirListing object.
func NewQDirListing(path string) *QDirListing {
	path_ms := struct_miqt_string{}
	path_ms.data = CString(path)
	path_ms.len = size_t(len(path))
	defer free(unsafe.Pointer(path_ms.data))
	g := newQDirListing(QDirListing_new(path_ms))
	g.isSubclass = true
	return g
}

// NewQDirListing2 constructs a new QDirListing object.
func NewQDirListing2(path string, nameFilters []string) *QDirListing {
	path_ms := struct_miqt_string{}
	path_ms.data = CString(path)
	path_ms.len = size_t(len(path))
	defer free(unsafe.Pointer(path_ms.data))
	nameFilters_CArray := (*[0xffff]struct_miqt_string)(malloc(size_t(int(unsafe.Sizeof(struct_miqt_string{})) * len(nameFilters))))
	defer free(unsafe.Pointer(nameFilters_CArray))
	for i := range nameFilters {
		nameFilters_i_ms := struct_miqt_string{}
		nameFilters_i_ms.data = CString(nameFilters[i])
		nameFilters_i_ms.len = size_t(len(nameFilters[i]))
		defer free(unsafe.Pointer(nameFilters_i_ms.data))
		nameFilters_CArray[i] = nameFilters_i_ms
	}
	nameFilters_ma := struct_miqt_array{len: size_t(len(nameFilters)), data: unsafe.Pointer(nameFilters_CArray)}
	g := newQDirListing(QDirListing_new2(path_ms, nameFilters_ma))
	g.isSubclass = true
	return g
}

// NewQDirListing3 constructs a new QDirListing object.
func NewQDirListing3(path string, flags IteratorFlags) *QDirListing {
	path_ms := struct_miqt_string{}
	path_ms.data = CString(path)
	path_ms.len = size_t(len(path))
	defer free(unsafe.Pointer(path_ms.data))
	g := newQDirListing(QDirListing_new3(path_ms, flags))
	g.isSubclass = true
	return g
}

// NewQDirListing4 constructs a new QDirListing object.
func NewQDirListing4(path string, nameFilters []string, flags IteratorFlags) *QDirListing {
	path_ms := struct_miqt_string{}
	path_ms.data = CString(path)
	path_ms.len = size_t(len(path))
	defer free(unsafe.Pointer(path_ms.data))
	nameFilters_CArray := (*[0xffff]struct_miqt_string)(malloc(size_t(int(unsafe.Sizeof(struct_miqt_string{})) * len(nameFilters))))
	defer free(unsafe.Pointer(nameFilters_CArray))
	for i := range nameFilters {
		nameFilters_i_ms := struct_miqt_string{}
		nameFilters_i_ms.data = CString(nameFilters[i])
		nameFilters_i_ms.len = size_t(len(nameFilters[i]))
		defer free(unsafe.Pointer(nameFilters_i_ms.data))
		nameFilters_CArray[i] = nameFilters_i_ms
	}
	nameFilters_ma := struct_miqt_array{len: size_t(len(nameFilters)), data: unsafe.Pointer(nameFilters_CArray)}
	g := newQDirListing(QDirListing_new4(path_ms, nameFilters_ma, flags))
	g.isSubclass = true
	return g
}

func (this *QDirListing) Swap(other *QDirListing) {
	QDirListing_Swap(this.h, other.cPointer())
}

func (this *QDirListing) IteratorPath() string {
	var _ms struct_miqt_string = QDirListing_IteratorPath(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing) IteratorFlags() IteratorFlags {
	xxxxxxxxx
}

func (this *QDirListing) NameFilters() []string {
	var _ma struct_miqt_array = QDirListing_NameFilters(this.h)
	_ret := make([]string, int(_ma.len))
	_outCast := (*[0xffff]struct_miqt_string)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		var _lv_ms struct_miqt_string = _outCast[i]
		_lv_ret := GoStringN(_lv_ms.data, int(int64(_lv_ms.len)))
		free(unsafe.Pointer(_lv_ms.data))
		_ret[i] = _lv_ret
	}
	return _ret
}

func (this *QDirListing) Begin() const_iterator {
	xxxxxxxxx
}

func (this *QDirListing) Cbegin() const_iterator {
	xxxxxxxxx
}

func (this *QDirListing) End() sentinel {
	xxxxxxxxx
}

func (this *QDirListing) Cend() sentinel {
	xxxxxxxxx
}

func (this *QDirListing) ConstBegin() const_iterator {
	xxxxxxxxx
}

func (this *QDirListing) ConstEnd() sentinel {
	xxxxxxxxx
}

type QDirListing__DirEntry struct {
	h          uintptr
	isSubclass bool
}

// NewQDirListing__DirEntry constructs a new QDirListing::DirEntry object.
func NewQDirListing__DirEntry(param1 *DirEntry) *QDirListing__DirEntry {
	g := newQDirListing__DirEntry(QDirListing__DirEntry_new(param1))
	g.isSubclass = true
	return g
}

// NewQDirListing__DirEntry2 constructs a new QDirListing::DirEntry object.
func NewQDirListing__DirEntry2() *QDirListing__DirEntry {
	g := newQDirListing__DirEntry(QDirListing__DirEntry_new2())
	g.isSubclass = true
	return g
}

func (this *QDirListing__DirEntry) FileName() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_FileName(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) BaseName() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_BaseName(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) CompleteBaseName() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_CompleteBaseName(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) Suffix() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_Suffix(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) BundleName() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_BundleName(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) CompleteSuffix() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_CompleteSuffix(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) FilePath() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_FilePath(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) IsDir() bool {
	return (bool)(QDirListing__DirEntry_IsDir(this.h))
}

func (this *QDirListing__DirEntry) IsFile() bool {
	return (bool)(QDirListing__DirEntry_IsFile(this.h))
}

func (this *QDirListing__DirEntry) IsSymLink() bool {
	return (bool)(QDirListing__DirEntry_IsSymLink(this.h))
}

func (this *QDirListing__DirEntry) Exists() bool {
	return (bool)(QDirListing__DirEntry_Exists(this.h))
}

func (this *QDirListing__DirEntry) IsHidden() bool {
	return (bool)(QDirListing__DirEntry_IsHidden(this.h))
}

func (this *QDirListing__DirEntry) IsReadable() bool {
	return (bool)(QDirListing__DirEntry_IsReadable(this.h))
}

func (this *QDirListing__DirEntry) IsWritable() bool {
	return (bool)(QDirListing__DirEntry_IsWritable(this.h))
}

func (this *QDirListing__DirEntry) IsExecutable() bool {
	return (bool)(QDirListing__DirEntry_IsExecutable(this.h))
}

func (this *QDirListing__DirEntry) FileInfo() *QFileInfo {
	_goptr := newQFileInfo(QDirListing__DirEntry_FileInfo(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDirListing__DirEntry) CanonicalFilePath() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_CanonicalFilePath(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) AbsoluteFilePath() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_AbsoluteFilePath(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) AbsolutePath() string {
	var _ms struct_miqt_string = QDirListing__DirEntry_AbsolutePath(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDirListing__DirEntry) Size() int64 {
	return (int64)(QDirListing__DirEntry_Size(this.h))
}

func (this *QDirListing__DirEntry) BirthTime(tz *QTimeZone) *QDateTime {
	_goptr := newQDateTime(QDirListing__DirEntry_BirthTime(this.h, tz.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDirListing__DirEntry) MetadataChangeTime(tz *QTimeZone) *QDateTime {
	_goptr := newQDateTime(QDirListing__DirEntry_MetadataChangeTime(this.h, tz.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDirListing__DirEntry) LastModified(tz *QTimeZone) *QDateTime {
	_goptr := newQDateTime(QDirListing__DirEntry_LastModified(this.h, tz.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDirListing__DirEntry) LastRead(tz *QTimeZone) *QDateTime {
	_goptr := newQDateTime(QDirListing__DirEntry_LastRead(this.h, tz.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDirListing__DirEntry) FileTime(typeVal QFileDevice__FileTime, tz *QTimeZone) *QDateTime {
	_goptr := newQDateTime(QDirListing__DirEntry_FileTime(this.h, (int)(typeVal), tz.cPointer()))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDirListing__DirEntry) OperatorAssign(param1 *DirEntry) {
	QDirListing__DirEntry_OperatorAssign(this.h, param1)
}

type QDirListing__sentinel struct {
	h          uintptr
	isSubclass bool
}

// NewQDirListing__sentinel constructs a new QDirListing::sentinel object.
func NewQDirListing__sentinel() *QDirListing__sentinel {
	g := newQDirListing__sentinel(QDirListing__sentinel_new())
	g.isSubclass = true
	return g
}

// NewQDirListing__sentinel2 constructs a new QDirListing::sentinel object.
func NewQDirListing__sentinel2(param1 *sentinel) *QDirListing__sentinel {
	g := newQDirListing__sentinel(QDirListing__sentinel_new2(param1))
	g.isSubclass = true
	return g
}

type QDirListing__const_iterator struct {
	h          uintptr
	isSubclass bool
}

// NewQDirListing__const_iterator constructs a new QDirListing::const_iterator object.
func NewQDirListing__const_iterator() *QDirListing__const_iterator {
	g := newQDirListing__const_iterator(QDirListing__const_iterator_new())
	g.isSubclass = true
	return g
}

func (this *QDirListing__const_iterator) OperatorMultiply() reference {
	xxxxxxxxx
}

func (this *QDirListing__const_iterator) OperatorMinusGreater() pointer {
	xxxxxxxxx
}

func (this *QDirListing__const_iterator) OperatorPlusPlus() *const_iterator {
	xxxxxxxxx
}

func (this *QDirListing__const_iterator) OperatorPlusPlusWithInt(param1 int) {
	QDirListing__const_iterator_OperatorPlusPlusWithInt(this.h, (int)(param1))
}
